.github/workflows/tag-build.yaml

```diff
@@ -1,10 +1,220 @@
+      - name: Set environment output values
+        id: set-variables
+        env:
+          GITHUB_EVENT_NAME: ${{ github.event_name}}
+          INPUT_VERSION: ${{ inputs.artifact-tag }}
+          SEMVER_REGEX: ^v?(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)\.(?:0|[1-9]\d*)(?:-(?:(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?:[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$
+        run: |
+          generate_version() {
+              # Example: v1.2.3-gen.4+g5678abcd
+              git describe --tags --match "v[[:digit:]]*.[[:digit:]]*.[[:digit:]]" | sed 's/\(.*\)-\(.*\)-\(.*\)/\1-gen.\2+\3/'
+          }
+          
+          get_output_vars() {
+              case "$GITHUB_EVENT_NAME" in
+                  "workflow_dispatch")
+                      # Case: workflow dispatch event. Pull most vars from inputs.
+                      echo "environment=build-stage"
+                      echo "version=$INPUT_VERSION"
+                      ;;
+                  "pull_request")
+                      echo "environment=build-stage"
+                      echo "version=$(generate_version "$GITHUB_EVENT_REF_NAME")"
+                      ;;
+                  "push")
+                      REF_TYPE=$(echo "$GITHUB_EVENT_REF" | cut -d'/' -f2)
+                      # Case: commit push event.
+                      if [ "$REF_TYPE" != "tags" ]; then
+                          echo "environment=build-stage"
+                          echo "version=$(generate_version "$GITHUB_EVENT_REF_NAME")"
+                          return
+                      fi
+          
+                      # Case: tag event with prerelease version.
+                      if [ "${GITHUB_EVENT_REF_NAME#*-}" != "$GITHUB_EVENT_REF_NAME" ]; then
+                          echo "environment=build-stage"
+                          echo "version=$GITHUB_EVENT_REF_NAME"
+                          return
+                      fi
+          
+                      # Case: tag event with release version. Only this
+                      # should go to prod.
+                      echo "environment=build-prod"
+                      echo "version=$GITHUB_EVENT_REF_NAME"
+                      ;;
+                  *)
+                      >&2 echo "Unknown GHA event $GITHUB_EVENT_NAME, failing" 
+                      exit 1
+                      ;;
+              esac
+          }
+          
+          get_output_vars >> "$GITHUB_OUTPUT"
+          
+          # Validate the semver
+          . "$GITHUB_OUTPUT"  # Load the variables into the current environment
+          (echo "$version" | grep -qP "$SEMVER_REGEX") || (echo "The artifact version $version is not a valid semver-coerced value"; exit 1)
+        
+          # Log the build details
+          echo "Built config:" | tee -a "$GITHUB_STEP_SUMMARY"
+          cat "$GITHUB_OUTPUT" | while read line; do
+            echo "* $line" | tee -a "$GITHUB_STEP_SUMMARY"
+          done
```

Comment:
> I would do this all in one block

I can do so but I don't really see the benefit. Is this to reduce the number of processes started and/or `open` calls by only invoking `tee` once?

> rather than writing a line iterator in shell

Why? Isn't using an iterator the opposite of [this above comment](https://github.com/gravitational/teleport-plugins/pull/983#discussion_r1470317343)?

Reply:
Not really; I'm specifically referring to a "line iterator". The unix toolkit has many commands for doing line iteration and processing - grep, sed, awk, etc. It is typically better to use these tools to operate over a sequence of lines rather than rolling your own iterators in shell; speed, correctness, grokability, etc are generally better. You want to modify every line of a file - reach for `sed`

For instance, there's a subtle bug in your loop - you should be using `read -r` to read a raw line and not interpret backslashes (https://www.shellcheck.net/wiki/SC2162 is the shellcheck warning on this).

As to the restructuring for a single `tee` - why do things N times when you only need to do it once?

Reply:
Updated pending push